zerver/views/auth.py

```diff
@@ -226,17 +231,38 @@ def login_or_register_remote_user(request: HttpRequest, remote_username: str,
     # or not they're using the mobile OTP flow or want a browser session.
     if mobile_flow_otp is not None:
         return finish_mobile_flow(request, user_profile, mobile_flow_otp)
+    elif desktop_flow_otp is not None:
+        assert realm is not None
+        return finish_desktop_flow(request, user_profile, realm, desktop_flow_otp)
 
     do_login(request, user_profile)
 
     redirect_to = get_safe_redirect_to(redirect_to, user_profile.realm.uri)
     return HttpResponseRedirect(redirect_to)
 
+def finish_desktop_flow(request: HttpRequest, user_profile: UserProfile,
+                        realm: Realm, otp: str) -> HttpResponse:
+    """
+    The desktop otp flow returns to the app (through a zulip:// redirect)
+    a token that allows obtaining (through log_into_subdomain) a logged in session
+    for the user account we authenticated in this flow.
+    The token can only be used once and within LOGIN_KEY_EXPIRATION_SECONDS
+    of being created, as nothing more powerful is needed for the desktop flow
+    and this minimizes the threat of one of these keys being leaked.
```

Comment:
"and this ensures the keys can only be used for completing this authentication attempt."

Reply:
Cool, changed that last one line in that commit.